File: apps/services/payments/src/observability.ts
import { Request, Response, NextFunction } from 'express';
import { Pool } from 'pg';
import { Registry, Counter, Histogram, Gauge, collectDefaultMetrics } from 'prom-client';
import { NodeSDK } from '@opentelemetry/sdk-node';
import { getNodeAutoInstrumentations } from '@opentelemetry/auto-instrumentations-node';
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-http';
import { OTLPMetricExporter } from '@opentelemetry/exporter-metrics-otlp-http';
import { PeriodicExportingMetricReader } from '@opentelemetry/sdk-metrics';

const registry = new Registry();
let metricsRegistered = false;
let sdkStarted = false;
let sdk: NodeSDK | undefined;

const HTTP_BUCKETS = [
  0.005,
  0.01,
  0.025,
  0.05,
  0.1,
  0.25,
  0.5,
  1,
  2.5,
  5,
  10,
];

const httpRequests = new Counter({
  name: 'apgms_http_requests_total',
  help: 'Total HTTP requests processed',
  labelNames: ['service', 'version', 'env', 'method', 'route', 'status'],
  registers: [registry],
});

const httpLatency = new Histogram({
  name: 'apgms_http_request_duration_seconds',
  help: 'HTTP request duration seconds',
  labelNames: ['service', 'version', 'env', 'method', 'route'],
  buckets: HTTP_BUCKETS,
  registers: [registry],
});

const httpInFlight = new Gauge({
  name: 'apgms_http_requests_in_flight',
  help: 'In-flight HTTP requests',
  labelNames: ['service', 'version', 'env'],
  registers: [registry],
});

const serviceMetadata = new Gauge({
  name: 'apgms_service_metadata',
  help: 'Static service metadata',
  labelNames: ['service', 'version', 'env'],
  registers: [registry],
});

const dbPoolConnections = new Gauge({
  name: 'apgms_db_pool_connections',
  help: 'Database pool usage by state',
  labelNames: ['service', 'version', 'env', 'pool', 'state'],
  registers: [registry],
});

const dlqGauge = new Gauge({
  name: 'apgms_dlq_messages',
  help: 'Messages currently buffered in DLQ',
  labelNames: ['service', 'version', 'env', 'queue'],
  registers: [registry],
});

const releaseFailures = new Counter({
  name: 'apgms_release_failures_total',
  help: 'Release pipeline failures recorded by the service',
  labelNames: ['service', 'version', 'env', 'stage'],
  registers: [registry],
});

export interface ObservabilityOptions {
  service: string;
  version: string;
  env: string;
  otlpEndpoint?: string;
}

export interface ObservabilityHandles {
  middleware: (req: Request, res: Response, next: NextFunction) => void;
  metricsHandler: (req: Request, res: Response) => Promise<void>;
  trackDbPool: (pool: Pool, name?: string) => void;
  recordReleaseFailure: (stage: string) => void;
  setDlqDepth: (queue: string, depth: number) => void;
  labels: { service: string; version: string; env: string };
}

function ensureOtelStarted(options: ObservabilityOptions): void {
  if (sdkStarted) {
    return;
  }
  const traceEndpoint =
    process.env.OTEL_EXPORTER_OTLP_TRACES_ENDPOINT ??
    process.env.OTEL_EXPORTER_OTLP_ENDPOINT ??
    options.otlpEndpoint ??
    'http://otel-collector:4318/v1/traces';
  const metricEndpoint =
    process.env.OTEL_EXPORTER_OTLP_METRICS_ENDPOINT ??
    process.env.OTEL_EXPORTER_OTLP_ENDPOINT ??
    options.otlpEndpoint ??
    'http://otel-collector:4318/v1/metrics';

  sdk = new NodeSDK({
    serviceName: options.service,
    traceExporter: new OTLPTraceExporter({ url: traceEndpoint }),
    metricReader: new PeriodicExportingMetricReader({
      exporter: new OTLPMetricExporter({ url: metricEndpoint }),
    }),
    instrumentations: [getNodeAutoInstrumentations()],
  });

  sdk
    .start()
    .then(() => {
      sdkStarted = true;
    })
    .catch((err) => {
      console.error('[observability] failed to start OpenTelemetry SDK', err);
    });

  const shutdown = async () => {
    if (!sdk) return;
    try {
      await sdk.shutdown();
    } catch (err) {
      console.error('[observability] failed to shutdown OpenTelemetry SDK', err);
    }
  };

  process.on('SIGTERM', shutdown);
  process.on('SIGINT', shutdown);
}

export function createObservability(options: ObservabilityOptions): ObservabilityHandles {
  if (!metricsRegistered) {
    registry.setDefaultLabels({
      service: options.service,
      version: options.version,
      env: options.env,
    });
    collectDefaultMetrics({ register: registry });
    metricsRegistered = true;
  }

  serviceMetadata.labels(options.service, options.version, options.env).set(1);
  ensureOtelStarted(options);

  const labelTuple = [options.service, options.version, options.env] as const;

  const middleware = (req: Request, res: Response, next: NextFunction) => {
    const start = process.hrtime.bigint();
    const method = req.method.toUpperCase();
    const routePath = req.route?.path ?? req.path ?? req.url;
    httpInFlight.labels(...labelTuple).inc();

    const done = (statusCode: number) => {
      const durationNs = Number(process.hrtime.bigint() - start);
      const durationSeconds = durationNs / 1e9;
      httpRequests.labels(...labelTuple, method, routePath, String(statusCode)).inc();
      httpLatency.labels(...labelTuple, method, routePath).observe(durationSeconds);
      httpInFlight.labels(...labelTuple).dec();
    };

    res.on('finish', () => done(res.statusCode));
    res.on('close', () => httpInFlight.labels(...labelTuple).dec());

    next();
  };

  const metricsHandler = async (_req: Request, res: Response) => {
    res.setHeader('Content-Type', registry.contentType);
    res.send(await registry.metrics());
  };

  const trackDbPool = (pool: Pool, name = 'default') => {
    const update = () => {
      dbPoolConnections.labels(...labelTuple, name, 'total').set(pool.totalCount);
      dbPoolConnections.labels(...labelTuple, name, 'idle').set(pool.idleCount);
      dbPoolConnections.labels(...labelTuple, name, 'waiting').set(pool.waitingCount);
      const active = Math.max(pool.totalCount - pool.idleCount, 0);
      dbPoolConnections.labels(...labelTuple, name, 'active').set(active);
    };
    update();
    pool.on('connect', update);
    pool.on('acquire', update);
    pool.on('release', update);
    pool.on('remove', update);
  };

  const recordReleaseFailure = (stage: string) => {
    releaseFailures.labels(...labelTuple, stage).inc();
  };

  const setDlqDepth = (queue: string, depth: number) => {
    dlqGauge.labels(...labelTuple, queue).set(depth);
  };

  return {
    middleware,
    metricsHandler,
    trackDbPool,
    recordReleaseFailure,
    setDlqDepth,
    labels: { service: options.service, version: options.version, env: options.env },
  };
}

File: apps/services/observability.py
"""Shared observability helpers for APGMS Python services."""
from __future__ import annotations

import os
import sys
import time
from dataclasses import dataclass
from typing import Any, Dict, Optional, Protocol

from prometheus_client import (
    CONTENT_TYPE_LATEST,
    Counter,
    Gauge,
    Histogram,
    generate_latest,
)

try:  # FastAPI optional import guard
    from fastapi import FastAPI, Response
    from starlette.requests import Request
except Exception:  # pragma: no cover - FastAPI not available in some contexts
    FastAPI = None  # type: ignore
    Response = None  # type: ignore
    Request = Any  # type: ignore


_HTTP_BUCKETS = (
    0.005,
    0.01,
    0.025,
    0.05,
    0.1,
    0.25,
    0.5,
    1,
    2.5,
    5,
    10,
)


HTTP_REQUESTS = Counter(
    "apgms_http_requests_total",
    "Total HTTP requests processed",
    ["service", "version", "env", "method", "route", "status"],
)
HTTP_LATENCY = Histogram(
    "apgms_http_request_duration_seconds",
    "HTTP request duration in seconds",
    ["service", "version", "env", "method", "route"],
    buckets=_HTTP_BUCKETS,
)
HTTP_IN_FLIGHT = Gauge(
    "apgms_http_requests_in_flight",
    "In-flight HTTP requests",
    ["service", "version", "env"],
)
SERVICE_METADATA = Gauge(
    "apgms_service_metadata",
    "Static service metadata",
    ["service", "version", "env"],
)
DB_CONNECTIONS = Gauge(
    "apgms_db_connections_active",
    "Active database connections",
    ["service", "version", "env"],
)
DLQ_DEPTH = Gauge(
    "apgms_dlq_messages",
    "Messages currently buffered in a dead-letter queue",
    ["service", "version", "env", "queue"],
)
RELEASE_FAILURES = Counter(
    "apgms_release_failures_total",
    "Release pipeline failures recorded by services",
    ["service", "version", "env", "stage"],
)


class SupportsClose(Protocol):
    def close(self) -> Any:
        ...


@dataclass
class InstrumentedConnection:
    _conn: SupportsClose
    _labels: Dict[str, str]
    _closed: bool = False

    def __post_init__(self) -> None:
        DB_CONNECTIONS.labels(**self._labels).inc()

    def __getattr__(self, item: str) -> Any:
        return getattr(self._conn, item)

    def close(self) -> Any:  # pragma: no cover - psycopg2 close returns None
        if not self._closed:
            DB_CONNECTIONS.labels(**self._labels).dec()
            self._closed = True
        return self._conn.close()

    def __enter__(self) -> "InstrumentedConnection":
        return self

    def __exit__(self, exc_type, exc, tb) -> Optional[bool]:
        self.close()
        return None


class Observability:
    """Helper to provide consistent metrics and tracing labels."""

    def __init__(self, service_name: str, *, version: str | None = None, env: str | None = None) -> None:
        self.service = service_name or os.getenv("SERVICE_NAME", "unknown-service")
        self.version = (
            version
            or os.getenv("SERVICE_VERSION")
            or os.getenv("VERSION")
            or os.getenv("APP_VERSION")
            or "dev"
        )
        self.env = env or os.getenv("SERVICE_ENV") or os.getenv("ENVIRONMENT") or "local"
        self._service_labels = {"service": self.service, "version": self.version, "env": self.env}
        SERVICE_METADATA.labels(**self._service_labels).set(1)

    @property
    def service_labels(self) -> Dict[str, str]:
        return dict(self._service_labels)

    def instrument_db_connection(self, conn: SupportsClose) -> InstrumentedConnection:
        return InstrumentedConnection(conn, self._service_labels)

    def set_dlq_depth(self, queue: str, depth: int) -> None:
        DLQ_DEPTH.labels(queue=queue, **self._service_labels).set(depth)

    def record_release_failure(self, stage: str) -> None:
        RELEASE_FAILURES.labels(stage=stage, **self._service_labels).inc()

    def install_metrics_endpoint(self, app: Any) -> None:
        if Response is None:
            return
        for route in getattr(app, "router", getattr(app, "routes", [])):  # pragma: no branch
            if getattr(route, "path", None) == "/metrics":
                return

        @app.get("/metrics")
        def _metrics() -> Response:
            payload = generate_latest()
            return Response(payload, media_type=CONTENT_TYPE_LATEST)

    def install_http_middleware(self, app: Any) -> None:
        if Request is None or FastAPI is None:
            return
        flag = "_apgms_http_metrics"
        if getattr(app.state, flag, False):
            return
        app.state.__setattr__(flag, True)

        @app.middleware("http")
        async def _metrics_middleware(request: Request, call_next):  # type: ignore[override]
            route_obj = request.scope.get("route")
            route_path = getattr(route_obj, "path", request.scope.get("path", request.url.path))
            method = request.method.upper()
            start = time.perf_counter()
            status_code = 500
            HTTP_IN_FLIGHT.labels(**self._service_labels).inc()
            try:
                response = await call_next(request)
                status_code = response.status_code
                return response
            finally:
                elapsed = time.perf_counter() - start
                HTTP_LATENCY.labels(route=route_path, method=method, **self._service_labels).observe(elapsed)
                HTTP_REQUESTS.labels(
                    route=route_path,
                    method=method,
                    status=str(status_code),
                    **self._service_labels,
                ).inc()
                HTTP_IN_FLIGHT.labels(**self._service_labels).dec()


def ensure_path_for_observability(current_file: str) -> None:
    """Append the shared services directory (where this module lives) to sys.path."""
    current = os.path.abspath(current_file)
    for _ in range(6):
        parent = os.path.dirname(current)
        if os.path.exists(os.path.join(parent, "observability.py")):
            if parent not in sys.path:
                sys.path.append(parent)
            return
        current = parent
